### PR TITLE
Steer to SetFeatureGateDuringTest on --feature-gates test usage.

### PR DESCRIPTION
**Pull request by benluddy**
#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by benluddy** [comment]
/cc @liggitt

Several integration tests can leak global feature overrides by passing --feature-gates to StartTestServer. Since you initially wrote `hack/verify-test-featuregates.sh` -- does something like this seem reasonable to you?

File: cmd/kube-apiserver/app/testing/testserver.go [hunk 1]
```diff
@@ -174,6 +174,11 @@ func StartTestServer(t Logger, instanceOptions *TestServerInstanceOptions, custo
 
 	fs := pflag.NewFlagSet("test", pflag.PanicOnError)
 
+	// Intercept feature gate flags to steer tests toward SetFeatureGateDuringTest.
```
Comment (liggitt) [inline cmd/kube-apiserver/app/testing/testserver.go:177]: is checking if `customFlags` contains `--feature-gates` or `--feature-gates=*` a simpler way of catching this

File: cmd/kube-apiserver/app/testing/testserver.go [hunk 2]
```diff
@@ -318,6 +323,14 @@ func StartTestServer(t Logger, instanceOptions *TestServerInstanceOptions, custo
 		return result, err
 	}
 
+	featuregatefs.VisitAll(func(flag *pflag.Flag) {
+		if !flag.Changed {
+			return
+		}
+
+		t.Fatalf("Flag %q mutates global state and should not be used in integration tests. Prefer SetFeatureGateDuringTest from k8s.io/component-base/featuregate/testing.", flag.Name)
```
Comment (liggitt) [inline cmd/kube-apiserver/app/testing/testserver.go:331]: alternatively, could add a snapshot/restore function to featuregate/testing like this:

```go
func SnapshotFeatureGateState(tb testing.TB, gate featuregate.FeatureGate) func() {
	var cleanups []func()

	// Snapshot all gates so their individual values get restored
	for k, v := range gate.(featuregate.MutableFeatureGate).GetAll() {
		if k == "AllAlpha" || k == "AllBeta" {
			continue
		}
		if !v.LockToDefault {
			cleanups = append(cleanups, SetFeatureGateDuringTest(tb, gate, k, gate.Enabled(k)))
		}
	}

	return func() {
		for _, cleanup := range cleanups {
			cleanup()
		}
	}
}
```

invoke that in this function:
```
restoreFeatureGates := featuregatetesting.SnapshotFeatureGateState(...)
```

and call `restoreFeatureGates()` from the `tearDown` function

**Pull request closed without merging**